### PR TITLE
Add children to needs

### PR DESCRIPTION
**Pull request by jbar485**
This PR resolves issue #180

### CODE REVIEW

**Comment by benjaminwood** [comment]
As it turns out, there are some specs to fix. Take a look and see what you can do. From a quick look, I believe we need to ensure that the `needs` factory builds one associate `child` record (which is now required for the need to be valid). See if the `users` factory has a similar setup because it requires one office assignment.

File: config/environments/test.rb
Comment (benjaminwood) [inline config/environments/test.rb:9]: @cidsinga can you undo this?

**Comment by dinjas** [comment]
This is getting pretty close. Specs pass, but there is a React-related error preventing the children fields for the needs form from rendering. I believe it is the same error that is causing the spec failures.

**Comment by dinjas** [comment]
*specs pass aside from afore-mentioned error that is causing two failures 😄

File: app/models/child.rb
Comment (benjaminwood) [inline app/models/child.rb:2]: Has anyone tested the behavior of re-using `number_of_children` here? For example, if an existing need was created with `number_of_children` set (before this is merged), are we at risk of it being cleared because there are in fact no associated Child records?

**Review by benjaminwood — CHANGES_REQUESTED** [review]
We need a strategy for existing needs that have `number_of_children` & `age_ranges`.

Perhaps that was the intent with the counter_cache? Anyways, let's think about this from the perspective of reporting on existing needs and new ones once this is deployed. I think reporting is a missing piece here.

**Comment by dinjas** [comment]
@benjaminwood - good call out. It looks like `number_of_children` wont get wiped out unless a child is added or removed. However - all existing Needs will become invalid once this is deployed, since they have no children. Wondering if we shouldnt attempt to create Child records for existing needs. Problem with that is that we don't know the gender or exact age of the child.

I have not looked at reporting yet, but will do that next.

**Comment by mireie** [comment]
Adding the new test has allowed all checks to pass.

**Comment by benjaminwood** [comment]
I did a bit of testing on staging and I was able to remove the 1 (and only) child from a need after creation. This should have been prevented, but it was not.

Could you test this locally to see if you can reproduce it? See if you can resolve it?

**Comment by mireie** [comment]
I'm able to recreate this issue and I think I've figured out why it's happening. When removing the final child and attempting to save a need this is the process:

- User updates a need and removes all children
- At time of save, @need currently has children associated and will pass validation
- When @need.save! is called the children records are destroyed
- @need is now invalid, but the child record(s) were already destroyed. @need will not update but the damage is done.

Looking into how to prevent `.save!` if it will remove all children.

**Comment by mireie** [comment]
> I did a bit of testing on staging and I was able to remove the 1 (and only) child from a need after creation. This should have been prevented, but it was not.
> 
> Could you test this locally to see if you can reproduce it? See if you can resolve it?

This should be resolved now. Added specs for new code as well.

File: app/models/need.rb
Comment (benjaminwood) [inline app/models/need.rb:54]: Do we intend to use this somewhere outside of specs?

File: app/controllers/needs_controller.rb
Comment (benjaminwood) [inline app/controllers/needs_controller.rb:62]: @mireie did you investigate why the model-level validation isn't preventing all the children from being removed? I presume there is something wrong with it, but I'm not sure what it is.

This fix does indeed resolve the symptom I experienced on staging, but I think a deeper fix should be sought out. Happy to look into it with you sometime.

File: app/models/need.rb
Comment (mireie) [reply app/models/need.rb:54]: Yes, this is used throughout the application. Currently, `number_of_children` is used to return the `number_of_children` property of a need. With this branch and the addition of children objects that belong to need adding this method allows the previous functionality to work like before but will instead return the total number of children objects instead of the legacy `number_of_children` property of the need.

File: app/controllers/needs_controller.rb
Comment (mireie) [reply app/controllers/needs_controller.rb:62]: I [did look into this](https://github.com/OfficeMomsandDads/scheduler/pull/446#issuecomment-962105870) and was not able to figure out why. I'd be happy to review.

Comment (mireie) [reply app/controllers/needs_controller.rb:62]: I still haven't figured out how to verify the model-level validation so I instead updated the React component to hide the delete button if there is only one child and added a check to `destroyChild` to only execute if there is more than one child.

**Review by benjaminwood — APPROVED** [review]
Excellent! Great job.

This (and other approved PR's) need some QA on staging, so let's wait for merge until that is done.

**Comment by benjaminwood** [comment]
@mireie looks like we have some merge conflicts now that other PR's have been merged. Can you resolve those?

**Comment by mireie** [comment]
Resolved!

**Pull request merged**
